exotations/solvers/exotica_ddp_solver/src/analytic_ddp_solver.cpp

```diff
@@ -0,0 +1,112 @@
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of  nor the names of its contributors may be used to
+//    endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include <exotica_ddp_solver/analytic_ddp_solver.h>
+
+REGISTER_MOTIONSOLVER_TYPE("AnalyticDDPSolver", exotica::AnalyticDDPSolver)
+
+namespace exotica
+{
+AnalyticDDPSolver::AnalyticDDPSolver() = default;
+AnalyticDDPSolver::~AnalyticDDPSolver() = default;
+
+void AnalyticDDPSolver::Instantiate(const AnalyticDDPSolverInitializer& init)
+{
+    parameters_ = init;
+    base_parameters_ = parameters_;
+}
+
+void AnalyticDDPSolver::BackwardPass()
+{
+    constexpr double min_clamp_ = -1e10;
+    constexpr double max_clamp_ = 1e10;
+    const int T = prob_->get_T();
+    const int NX = prob_->get_num_positions();
+    const int NU = prob_->get_num_controls();
+    const int N = NX + prob_->get_num_velocities();
+    const double dt_squared = dynamics_solver_->get_dt() * dynamics_solver_->get_dt();
+
+    Eigen::MatrixXd Qx, Qu, Qxx, Quu, Qux, Quu_inv, Vxx;
+    Eigen::VectorXd Vx;
+
+    // V = prob_->GetStateCost(T - 1);
+    Vx = prob_->GetStateCostJacobian(T - 1);
+    Vxx = prob_->GetStateCostHessian(T - 1);
+
+    // concatenation axis for tensor products
+    Eigen::array<Eigen::IndexPair<int>, 1> dims = {Eigen::IndexPair<int>(1, 0)};
+
+    for (int t = T - 2; t > 0; t--)
+    {
+        Eigen::VectorXd x = prob_->get_X(t), u = prob_->get_U(t);
+        Eigen::MatrixXd fx = dynamics_solver_->fx(x, u),
+                        fu = dynamics_solver_->fu(x, u);
+
+        fx = fx * dynamics_solver_->get_dt() + Eigen::MatrixXd::Identity(fx.rows(), fx.cols());
+        fu = fu * dynamics_solver_->get_dt();
+
+        // Q = prob_->GetStateCost(t) + prob_->GetControlCost(t) + V; // l + v
+        Qx = prob_->GetStateCostJacobian(t) + fx.transpose() * Vx;  // lx + fx @ Vx
+        Qu = prob_->GetControlCostJacobian(t) + fu.transpose() * Vx;
+        Qxx = prob_->GetStateCostHessian(t) + fx.transpose() * Vxx * fx;
+        Quu = prob_->GetControlCostHessian() + fu.transpose() * Vxx * fu;
```

Comment:
You could use a temporary here for fx^T * Vxx - and then only multiplies fx or fu as required. That can help (Croccodyl does it this way).

Reply:
But one is `fx * Vxx` and the other - `fu * Vxx`? You could potentially use a temp for the second one..
Also, does't look like Croccodyl is doing this (at least in python):
https://gepgitlab.laas.fr/loco-3d/crocoddyl/blob/master/crocoddyl/ddp.py